extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java

```diff
@@ -144,6 +151,71 @@ public Long terminatePartial(RamAccounting ramAccounting, HllState state) {
+            case DoubleType.ID:
+            case FloatType.ID:
+                return new SortedNumericDocValueAggregator<>(
+                    fieldTypes.get(0).name(),
+                    (ramAccounting, memoryManager, minNodeVersion) -> {
+                        var state = new HllState(dataType, minNodeVersion.onOrAfter(Version.V_4_1_0));
+                        Integer precision = optionalParams.length == 1 ? (Integer) optionalParams[0].value() : HyperLogLogPlusPlus.DEFAULT_PRECISION;
+                        return initIfNeeded(state, memoryManager, precision);
+                    },
+                    (values, state) -> {
+                        // Murmur3Hash.Long and Murmur3Hash.Double in regular aggregation call mix64 of doubleToLongBits(double value).
+                        // Equivalent of that in context of DocValue is
+                        // mix64(doubleToLongBits(decoded))
+                        // => mix64(doubleToLongBits(NumericUtils.sortableLongToDouble(values.nextValue())))
+                        // => mix64(doubleToLongBits(Double.longBitsToDouble(sortableDoubleBits(values.nextValue()))))
+                        // => mix64(sortableDoubleBits(values.nextValue()))
+
+                        var hash = BitMixer.mix64(NumericUtils.sortableDoubleBits(values.nextValue()));
+                        state.addHash(hash);
+                    }
+                );
+            case IpType.ID:
+            case StringType.ID:
+                return new BinaryDocValueAggregator<>(
+                    fieldTypes.get(0).name(),
+                    (ramAccounting, memoryManager, minNodeVersion) -> {
+                        var state = new HllState(dataType, minNodeVersion.onOrAfter(Version.V_4_1_0));
+                        Integer precision = optionalParams.length == 1 ? (Integer) optionalParams[0].value() : HyperLogLogPlusPlus.DEFAULT_PRECISION;
+                        return initIfNeeded(state, memoryManager, precision);
+                    },
+                    (values, state) -> {
+                        BytesRef ref = values.nextValue();
+                        // Important: using ref.bytes for getting utf-8 bytearray is wrong,
+                        // ref.bytes and DataTypes.STRING.implicitCast(ref).getBytes(StandardCharsets.UTF_8)
+                        // produce different arrays with same beginnings but ref.bytes has some "unused" bytes at the end.
+                        // BytesRef has a property 'length' providing length of used bytes.
+
+                        // ref.bytes is used below since hash methods has a 'length' argument to control real length of the array
+                        // and ref.length is used to ignore unused bytes.
+                        var hash = state.isAllOn4_1() ?
+                            MurmurHash3.hash64(ref.bytes, ref.length)
+                            : MurmurHash3.hash128(ref.bytes, 0, ref.length, 0, Murmur3Hash.Bytes.INSTANCE.hash).h1;
```

Comment:
The `offset` of the `BytesRef` could also be `> 0` in which case the logic here would probably be wrong.

Reply:
Using `Murmur3Hash.Bytes.INSTANCE.hash` here will not be thread-safe. The hash128 has to be a local instance that can be created within this `BinaryDocValueAggregator` instance.

Reply:
Good catch, thanks!